src/serde.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Internal `serde` helpers.
+
+use core::fmt::Binary;
+use core::fmt::LowerHex;
+use core::marker::PhantomData;
+use core::mem;
+
+#[cfg(feature = "std")]
+use std::ffi::CString;
+
+use serde::de;
+use serde::Deserialize;
+use serde::Deserializer;
+use serde::Serializer;
+
+/// For defaulting a field to `true`.
+pub fn default_to_true() -> bool {
+    true
+}
+
+/// For skipping field serialization if it's set to `true`.
+///
+/// (Use with `default_to_true()`.
+pub fn skip_if_true(b: &bool) -> bool {
+    *b
+}
+
+/// For deserializing a `Vec<u8>` from either a string or a sequence of bytes.
+#[cfg(feature = "std")]
+pub fn de_bytestring<'de, D>(d: D) -> Result<Vec<u8>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    // CString "does the right thing" per serde's implementation.
+    Ok(CString::deserialize(d)?.into_bytes())
+}
+
+/// For serializing a `Vec<u8>` as a bytestring.
+#[cfg(feature = "std")]
+pub fn se_bytestring<S>(bytes: &Vec<u8>, s: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    s.serialize_bytes(bytes)
+}
+
+/// Helper for `de_radix`.
+pub struct Radix<T>(PhantomData<T>);
+
+macro_rules! impl_radix {
+    ($($ty:ident)*) => {$(
+        impl<'de> de::Visitor<'de> for Radix<$ty> {
+            type Value = $ty;
+
+            fn expecting(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                write!(f, "integer between 0 and {}", core::$ty::MAX)
+            }
+            fn visit_borrowed_str<E>(self, s: &'de str) -> Result<$ty, E>
+                where E: de::Error,
+            {
+                let int = if s.starts_with("0b") || s.starts_with("0B") {
+                    $ty::from_str_radix(&s[2..], 0b10)
+                }
+                else if s.starts_with("0o") || s.starts_with("0O") {
+                    $ty::from_str_radix(&s[2..], 0o10)
+                }
+                else if s.starts_with("0x") || s.starts_with("0X") {
+                    $ty::from_str_radix(&s[2..], 0x10)
+                } else {
+                    $ty::from_str_radix(s, 10)
```

Comment:
2, 8, 16 would be significantly more readable.

Reply:
Too clever huh?